### PR TITLE
Setting panel sizes in `theme()`

### LINKED ISSUE #5338: Panel-based sizing of figures
(Apologies if this has been raised and addressed before)

Despite all the marvels of ggplot2 and the many amazing extension packages, one thing that continues to consume a lot of time (or often is difficult to achieve at all) is the consistent sizing of figures for a publication. Specifically, I find that almost all the time my focus is on the relative sizing of _panels_ (rather than the entire area of a figure) across figures. A common scenario is, for example, that I have the same multi-panel result plots across multiple figures. Another common scenario is that I'd like to plot a panel in Figure 1 at twice the width of Figure 2, etc. Critically, in all of these situations, my focus is on the panels rather than the entire figure. 

I've searched across packages but I haven't found any easy way to create alignment in panel sizes across plots (I'd love to hear that I'm wrong!). Would it be possible, for example, to allow ggsave to save figures based on the panel size, or to provide other ways to easily express the dimensions of the panel itself. At the core of the issue that I'm trying to describe is that sizing seems to start by sizing the overall figure. The rest follows from that (incl. the panel size). But perhaps many researchers would find it more intuitive if sizing started with the panel? (plotting figures at 100% would then guarantee that all panels have proportional sizes as intended by the author). 

Sorry, if this is all already solved or known to be unsolvable. In any case, thank you for all the amazing work on ggplot!!!

### PR DESCRIPTION
**Pull request by teunbrand**
This PR aims to fix #5338.

Briefly, the layout object gets a new method that sets the panel sizes.

The gist is that you can give a vector of units to `theme(panel.widths)` or `theme(panel.heights)` that then applies to each panel. It overrules any aspect ratios set in the coord or theme.

Somewhat special behaviour occurs if you give a unit of length 1, but you have multiple panels. In that case, the total panel area is set, rather than the size of individual panels.

``` r
devtools::load_all("~/packages/ggplot2")
#> ℹ Loading ggplot2

p <- ggplot(mpg, aes(displ, hwy)) +
  geom_point() +
  facet_grid(~ drv)

# Square 2x2cm panels
p + theme(
  panel.widths = unit(c(2, 2, 2), "cm"),
  panel.heights = unit(2, "cm")
)
```

``` r
# Total width is 6cm, so every panel is somewhat narrower than 2cm
p + theme(
  panel.widths = unit(6, "cm"),
  panel.heights = unit(2, "cm")
)
```

<sup>Created on 2024-09-09 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

I'm not quite sure whether the layout is the correct place for the method. It can also belong to the facet class, for example.

### CODE REVIEW

**Comment by thomasp85** [comment]
With dev patchwork it now works in the sense that the dimensions gets adopted from the panel with the largest dimensions (if set in absolute units)

There is really no way to accommodate conflicting sizes while keeping the alignment model of patchwork so something got to give

File: R/layout.R [hunk 1]
```diff
@@ -297,10 +298,56 @@ Layout <- ggproto("Layout", NULL,
...
+    if (is.null(new_widths) && is.null(new_heights)) {
+      return(table)
+    }
+
+    if (isTRUE(table$respect)) {
+      args <- !c(is.null(new_widths), is.null(new_heights))
+      args <- c("panel.widths", "panel.heights")[args]
+      cli::cli_warn(
+        "Aspect ratios are overruled by {.arg {args}} theme element{?s}."
+      )
```
[the remark this answers is not in the document]
Comment (teunbrand) [reply R/layout.R:317]: Gotcha

**Comment by thomasp85** [comment]
I _think_ this all belongs in `Facet` and not `Layout`

**Comment by teunbrand** [comment]
The nice thing about making this a Facet method is that extensions can implement their own if their gtable structure differs

**Review by thomasp85 — APPROVED** [review]
LGTM

**Pull request merged**